sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +969,304 @@ cdef class MSE(RegressionCriterion):
 
+cdef class MAE(RegressionCriterion):
+    """Mean absolute error impurity criterion"""
+    def __dealloc__(self):
+        """Destructor."""
+
+        free(self.node_medians)
+
+    cdef np.ndarray left_child_heaps
+    cdef np.ndarray right_child_heaps
+    cdef double* node_medians
+
+    def __cinit__(self, SIZE_t n_outputs):
+        """Initialize parameters for this criterion.
+
+        Parameters
+        ----------
+        n_outputs: SIZE_t
+            The number of targets to be predicted
+        """
+
+        # Default values
+        self.y = NULL
+        self.y_stride = 0
+        self.sample_weight = NULL
+
+        self.samples = NULL
+        self.start = 0
+        self.pos = 0
+        self.end = 0
+
+        self.n_outputs = n_outputs
+        self.n_node_samples = 0
+        self.weighted_n_node_samples = 0.0
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = 0.0
+
+        # Allocate accumulators. Make sure they are NULL, not uninitialized,
+        # before an exception can be raised (which triggers __dealloc__).
+        self.node_medians = NULL
+
+        # Allocate memory for the accumulators
+        self.node_medians = <double*> calloc(n_outputs, sizeof(double))
+
+        if (self.node_medians == NULL):
+            raise MemoryError()
+
+        self.left_child_heaps = np.empty(n_outputs, dtype='object')
+        self.right_child_heaps = np.empty(n_outputs, dtype='object')
+
+    cdef void init(self, DOUBLE_t* y, SIZE_t y_stride, DOUBLE_t* sample_weight,
+                   double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                   SIZE_t end) nogil:
+        """Initialize the criterion at node samples[start:end] and
+           children samples[start:start] and samples[start:end]."""
+
+        # Initialize fields
+        self.y = y
+        self.y_stride = y_stride
+        self.sample_weight = sample_weight
+        self.samples = samples
+        self.start = start
+        self.end = end
+        self.n_node_samples = end - start
+        self.weighted_n_samples = weighted_n_samples
+        self.weighted_n_node_samples = 0.
+
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef SIZE_t k
+        cdef DOUBLE_t y_ik
+        cdef DOUBLE_t w_y_ik
+        cdef DOUBLE_t w = 1.0
+
+        # Fill accumulators with MedianHeaps
+        with gil:
+            for k in range(self.n_outputs):
+                self.left_child_heaps[k] = MedianHeap(self.n_node_samples)
+                self.right_child_heaps[k] = MedianHeap(self.n_node_samples)
+
+        cdef void** left_child_heaps = <void**> self.left_child_heaps.data
+        cdef void** right_child_heaps = <void**> self.right_child_heaps.data
+
+        for p in range(start, end):
+            i = samples[p]
+
+            if sample_weight != NULL:
+                w = sample_weight[i]
+
+            for k in range(self.n_outputs):
+                y_ik = y[i * y_stride + k]
+                w_y_ik = w * y_ik
```

Comment:
@glouppe ironically, your counterexample works with what we're doing 😆 Our thinking was that the median doesn't actually matter that much, as it is but an intermediate step to the impurity. On your counter example, the MAE for [1,2,3,4,5] is [(3-1) + (3-2) + (3-3) + (4-3) +(5/3)]  / 5 = (2+1+1+2)/5 = 6/5.

Given the values [1,2,3,4,5] and weights [2,2,2,2,2], we get the product [2 4 6 8 10] (median of 6). We then subtract this median from the other products, and critically divide by the total weights. So in this case, we calculate [(6-2) + (6-4) + (6-6) + (8-6) + (10-6)] / 10 = 4+2+2+4 / 10 = 12/10 = 6/5.

However, just because your counter-example doesn't work doesn't mean you aren't correct. There are issues with the calculation scheme above, as demonstrated by the trying out the values [1,1,1] and the weights [1,2,1] or the values [1,5,10] and the weights [1,2,3].